File: resources/test/language_samples/sample.pas
program SamplePascalProgram;

{$mode objfpc}{$H+}
{$APPTYPE CONSOLE}

{ Sample Pascal/Delphi/Object Pascal program demonstrating various language features }

uses
  SysUtils, Classes, Math, Variants, DateUtils;

type
  { Forward declarations }
  TAnimal = class;
  
  { Custom exception class }
  EInvalidAgeException = class(Exception);
  
  { Abstract base class demonstrating OOP concepts }
  TAnimal = class abstract
  private
    FName: string;
    FAge: Integer;
    FSpecies: string;
  protected
    procedure SetAge(const Value: Integer); virtual;
  public
    constructor Create(const AName, ASpecies: string; AAge: Integer);
    destructor Destroy; override;
    
    { Abstract method - must be implemented in derived classes }
    function MakeSound: string; virtual; abstract;
    
    { Virtual method with default implementation }
    function GetDescription: string; virtual;
    
    { Properties }
    property Name: string read FName write FName;
    property Age: Integer read FAge write SetAge;
    property Species: string read FSpecies;
  end;
  
  { Concrete implementation }
  TDog = class(TAnimal)
  public
    constructor Create(const AName: string; AAge: Integer);
    function MakeSound: string; override;
    function GetDescription: string; override;
    
    { Additional methods }
    procedure Fetch(const Item: string);
  end;
  
  { Another concrete implementation }
  TCat = class(TAnimal)
  private
    FIsIndoor: Boolean;
  public
    constructor Create(const AName: string; AAge: Integer; AIsIndoor: Boolean);
    function MakeSound: string; override;
    procedure Purr;
    
    property IsIndoor: Boolean read FIsIndoor write FIsIndoor;
  end;
  
  { Generic list demonstration }
  {$IFDEF FPC}
  generic TAnimalList<T: TAnimal> = class(TList)
  private
    function GetItem(Index: Integer): T;
    procedure SetItem(Index: Integer; const Value: T);
  public
    property Items[Index: Integer]: T read GetItem write SetItem; default;
    procedure AddAnimal(Animal: T);
  end;
  {$ENDIF}
  
  { Record type with methods (modern Pascal feature) }
  TPoint = record
    X, Y: Double;
    
    class function Create(AX, AY: Double): TPoint; static;
    function Distance(const Other: TPoint): Double;
    function ToString: string;
  end;
  
  { Enumeration with explicit values }
  TColor = (clRed = 1, clGreen = 2, clBlue = 4, clYellow = 8);
  TColors = set of TColor;
  
  { Variant record (union) }
  TShape = record
    case ShapeType: (stCircle, stRectangle, stTriangle) of
      stCircle: (Radius: Double);
      stRectangle: (Width, Height: Double);
      stTriangle: (Base, TriangleHeight: Double);
  end;

{ Global variables }
var
  GlobalCounter: Integer = 0;

{ Function with multiple parameters and default values }
function CalculateArea(const Shape: TShape; Precision: Integer = 2): Double;
begin
  case Shape.ShapeType of
    stCircle:
      Result := Pi * Sqr(Shape.Radius);
    stRectangle:
      Result := Shape.Width * Shape.Height;
    stTriangle:
      Result := 0.5 * Shape.Base * Shape.TriangleHeight;
  else
    Result := 0.0;
  end;
  
  Result := RoundTo(Result, -Precision);
end;

{ Procedure with var and const parameters }
procedure ProcessNumbers(const Input: array of Integer; var Sum, Average: Double; out Count: Integer);
var
  I: Integer;
begin
  Sum := 0;
  Count := Length(Input);
  
  for I := Low(Input) to High(Input) do
    Sum := Sum + Input[I];
    
  if Count > 0 then
    Average := Sum / Count
  else
    Average := 0;
end;

{ String manipulation example }
function FormatPersonInfo(const FirstName, LastName: string; Age: Integer): string;
begin
  Result := Format('%s, %s (Age: %d)', [LastName, FirstName, Age]);
end;

{ File I/O example }
procedure WriteToFile(const FileName, Content: string);
var
  FileStream: TFileStream;
  StringStream: TStringStream;
begin
  try
    FileStream := TFileStream.Create(FileName, fmCreate);
    try
      StringStream := TStringStream.Create(Content);
      try
        FileStream.CopyFrom(StringStream, 0);
      finally
        StringStream.Free;
      end;
    finally
      FileStream.Free;
    end;
  except
    on E: Exception do
      WriteLn('Error writing to file: ', E.Message);
  end;
end;

{ Exception handling example }
function SafeDivision(A, B: Double): Double;
begin
  try
    if B = 0 then
      raise EDivByZero.Create('Division by zero is not allowed');
    Result := A / B;
  except
    on E: EDivByZero do
    begin
      WriteLn('Division error: ', E.Message);
      Result := 0;
    end;
    on E: Exception do
    begin
      WriteLn('Unexpected error: ', E.Message);
      Result := 0;
    end;
  end;
end;

{ Implementation of TAnimal methods }
constructor TAnimal.Create(const AName, ASpecies: string; AAge: Integer);
begin
  inherited Create;
  FName := AName;
  FSpecies := ASpecies;
  SetAge(AAge);
end;

destructor TAnimal.Destroy;
begin
  WriteLn(Format('Animal %s has been destroyed', [FName]));
  inherited Destroy;
end;

procedure TAnimal.SetAge(const Value: Integer);
begin
  if Value < 0 then
    raise EInvalidAgeException.Create('Age cannot be negative');
  FAge := Value;
end;

function TAnimal.GetDescription: string;
begin
  Result := Format('%s is a %d-year-old %s', [FName, FAge, FSpecies]);
end;

{ Implementation of TDog methods }
constructor TDog.Create(const AName: string; AAge: Integer);
begin
  inherited Create(AName, 'Dog', AAge);
end;

function TDog.MakeSound: string;
begin
  Result := 'Woof! Woof!';
end;

function TDog.GetDescription: string;
begin
  Result := inherited GetDescription + ' and loves to play fetch!';
end;

procedure TDog.Fetch(const Item: string);
begin
  WriteLn(Format('%s is fetching the %s!', [Name, Item]));
end;

{ Implementation of TCat methods }
constructor TCat.Create(const AName: string; AAge: Integer; AIsIndoor: Boolean);
begin
  inherited Create(AName, 'Cat', AAge);
  FIsIndoor := AIsIndoor;
end;

function TCat.MakeSound: string;
begin
  Result := 'Meow! Purr...';
end;

procedure TCat.Purr;
begin
  WriteLn(Format('%s is purring contentedly...', [Name]));
end;

{ Implementation of TPoint methods }
class function TPoint.Create(AX, AY: Double): TPoint;
begin
  Result.X := AX;
  Result.Y := AY;
end;

function TPoint.Distance(const Other: TPoint): Double;
begin
  Result := Sqrt(Sqr(X - Other.X) + Sqr(Y - Other.Y));
end;

function TPoint.ToString: string;
begin
  Result := Format('(%.2f, %.2f)', [X, Y]);
end;

{ Main program }
var
  Dog: TDog;
  Cat: TCat;
  Numbers: array[1..5] of Integer = (10, 20, 30, 40, 50);
  Sum, Average: Double;
  Count: Integer;
  Point1, Point2: TPoint;
  Shape: TShape;
  Area: Double;
  Colors: TColors;
  I: Integer;

begin
  try
    WriteLn('=== Pascal/Delphi Sample Program ===');
    WriteLn;
    
    { Object-oriented programming example }
    WriteLn('--- Object-Oriented Programming ---');
    Dog := TDog.Create('Buddy', 3);
    Cat := TCat.Create('Whiskers', 2, True);
    
    try
      WriteLn(Dog.GetDescription);
      WriteLn('Dog says: ', Dog.MakeSound);
      Dog.Fetch('ball');
      WriteLn;
      
      WriteLn(Cat.GetDescription);
      WriteLn('Cat says: ', Cat.MakeSound);
      Cat.Purr;
      WriteLn;
    finally
      Dog.Free;
      Cat.Free;
    end;
    
    { Array processing }
    WriteLn('--- Array Processing ---');
    ProcessNumbers(Numbers, Sum, Average, Count);
    WriteLn(Format('Numbers processed: %d, Sum: %.2f, Average: %.2f', [Count, Sum, Average]));
    WriteLn;
    
    { Record and geometry calculations }
    WriteLn('--- Geometry Calculations ---');
    Point1 := TPoint.Create(0, 0);
    Point2 := TPoint.Create(3, 4);
    WriteLn(Format('Distance between %s and %s: %.2f', 
                   [Point1.ToString, Point2.ToString, Point1.Distance(Point2)]));
    
    { Variant record example }
    Shape.ShapeType := stCircle;
    Shape.Radius := 5.0;
    Area := CalculateArea(Shape);
    WriteLn(Format('Circle area (radius %.1f): %.2f', [Shape.Radius, Area]));
    
    Shape.ShapeType := stRectangle;
    Shape.Width := 4.0;
    Shape.Height := 6.0;
    Area := CalculateArea(Shape);
    WriteLn(Format('Rectangle area (%.1f x %.1f): %.2f', [Shape.Width, Shape.Height, Area]));
    WriteLn;
    
    { Set operations }
    WriteLn('--- Set Operations ---');
    Colors := [clRed, clBlue];
    if clRed in Colors then
      WriteLn('Red is in the color set');
    Colors := Colors + [clGreen];
    WriteLn(Format('Color set now contains %d colors', [Ord(clRed in Colors) + 
                    Ord(clGreen in Colors) + Ord(clBlue in Colors) + Ord(clYellow in Colors)]));
    WriteLn;
    
    { Exception handling }
    WriteLn('--- Exception Handling ---');
    WriteLn(Format('Safe division 10/2 = %.2f', [SafeDivision(10, 2)]));
    WriteLn(Format('Safe division 10/0 = %.2f', [SafeDivision(10, 0)]));
    WriteLn;
    
    { Loop examples }
    WriteLn('--- Loop Examples ---');
    Write('For loop: ');
    for I := 1 to 5 do
      Write(I, ' ');
    WriteLn;
    
    Write('While loop countdown: ');
    I := 5;
    while I > 0 do
    begin
      Write(I, ' ');
      Dec(I);
    end;
    WriteLn;
    
    { String operations }
    WriteLn('--- String Operations ---');
    WriteLn(FormatPersonInfo('John', 'Doe', 30));
    WriteLn(FormatPersonInfo('Jane', 'Smith', 25));
    
    WriteLn;
    WriteLn('Program completed successfully!');
    
  except
    on E: Exception do
    begin
      WriteLn('An error occurred: ', E.Message);
      ExitCode := 1;
    end;
  end;
  
  {$IFDEF MSWINDOWS}
  WriteLn('Press Enter to exit...');
  ReadLn;
  {$ENDIF}
end. 

File: resources/test/language_samples/sample.scala
// Scala Sample Program - Demonstrating functional and object-oriented programming
// This program showcases various Scala language features including:
// - Object-oriented programming with classes and traits
// - Functional programming with higher-order functions
// - Pattern matching and case classes
// - Collections and immutable data structures
// - Concurrent programming with Futures

import scala.concurrent.{Future, ExecutionContext}
import scala.util.{Try, Success, Failure, Random}
import scala.collection.mutable
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

// Implicit execution context for Future operations
implicit val ec: ExecutionContext = ExecutionContext.global

// Sealed trait for algebraic data types
sealed trait Shape {
  def area: Double
  def perimeter: Double
}

// Case classes for different shapes
case class Circle(radius: Double) extends Shape {
  def area: Double = math.Pi * radius * radius
  def perimeter: Double = 2 * math.Pi * radius
  
  override def toString: String = s"Circle(radius=$radius)"
}

case class Rectangle(width: Double, height: Double) extends Shape {
  def area: Double = width * height
  def perimeter: Double = 2 * (width + height)
  
  override def toString: String = s"Rectangle(width=$width, height=$height)"
}

case class Triangle(base: Double, height: Double, side1: Double, side2: Double) extends Shape {
  def area: Double = 0.5 * base * height
  def perimeter: Double = base + side1 + side2
  
  override def toString: String = s"Triangle(base=$base, height=$height)"
}

// Trait for objects that can be measured
trait Measurable {
  def dimensions: Map[String, Double]
}

// Abstract class demonstrating inheritance
abstract class Animal(val name: String, val species: String) {
  def makeSound: String
  def description: String = s"$name is a $species"
  
  // Abstract method to be implemented by subclasses
  def habitat: String
}

// Concrete implementations
class Dog(name: String, val breed: String) extends Animal(name, "Canis lupus") {
  def makeSound: String = "Woof! Woof!"
  def habitat: String = "Domestic environment"
  
  def fetch(item: String): String = s"$name is fetching the $item!"
  
  override def description: String = s"${super.description} of breed $breed"
}

class Cat(name: String, val isIndoor: Boolean) extends Animal(name, "Felis catus") {
  def makeSound: String = "Meow! Purr..."
  def habitat: String = if (isIndoor) "Indoor" else "Indoor/Outdoor"
  
  def purr(): String = s"$name is purring contentedly..."
}

// Case class for immutable data
case class Person(firstName: String, lastName: String, age: Int, email: String) {
  def fullName: String = s"$firstName $lastName"
  def isAdult: Boolean = age >= 18
  
  // Copy method is automatically generated for case classes
  def withAge(newAge: Int): Person = this.copy(age = newAge)
}

// Companion object with factory methods
object Person {
  def apply(fullName: String, age: Int, email: String): Person = {
    val parts = fullName.split(" ", 2)
    Person(parts(0), parts.lift(1).getOrElse(""), age, email)
  }
  
  def fromEmail(email: String, age: Int): Person = {
    val username = email.split("@")(0)
    val parts = username.split("[._]")
    Person(parts(0).capitalize, parts.lift(1).map(_.capitalize).getOrElse(""), age, email)
  }
}

// Generic class demonstrating type parameters
class GenericContainer[T](private var value: T) {
  def get: T = value
  def set(newValue: T): Unit = value = newValue
  def transform[U](f: T => U): GenericContainer[U] = new GenericContainer(f(value))
  
  override def toString: String = s"Container($value)"
}

// Object with utility functions
object MathUtils {
  def fibonacci(n: Int): BigInt = {
    @annotation.tailrec
    def fibHelper(n: Int, prev: BigInt, curr: BigInt): BigInt = {
      if (n <= 0) prev
      else fibHelper(n - 1, curr, prev + curr)
    }
    fibHelper(n, 0, 1)
  }
  
  def factorial(n: Int): BigInt = {
    if (n <= 1) 1
    else (2 to n).map(BigInt(_)).product
  }
  
  def isPrime(n: Int): Boolean = {
    if (n <= 1) false
    else if (n <= 3) true
    else if (n % 2 == 0 || n % 3 == 0) false
    else {
      val sqrt = math.sqrt(n).toInt
      (5 to sqrt by 6).forall(i => n % i != 0 && n % (i + 2) != 0)
    }
  }
}

// Option and error handling
object FileProcessor {
  def readFile(filename: String): Option[String] = {
    Try {
      scala.io.Source.fromFile(filename).getLines().mkString("\n")
    }.toOption
  }
  
  def processContent(content: String): Either[String, List[String]] = {
    if (content.trim.isEmpty) {
      Left("Content is empty")
    } else {
      Right(content.split("\\n").toList.filter(_.nonEmpty))
    }
  }
}

// Future-based asynchronous operations
object AsyncOperations {
  def simulateNetworkCall(delay: Int): Future[String] = {
    Future {
      Thread.sleep(delay)
      s"Network response after ${delay}ms"
    }
  }
  
  def processDataAsync(data: List[Int]): Future[List[Int]] = {
    Future {
      data.map(_ * 2).filter(_ > 10)
    }
  }
}

// Pattern matching examples
object PatternMatchingExamples {
  def describeValue(value: Any): String = value match {
    case s: String if s.length > 10 => s"Long string: ${s.take(10)}..."
    case s: String => s"String: $s"
    case i: Int if i > 0 => s"Positive integer: $i"
    case i: Int if i < 0 => s"Negative integer: $i"
    case 0 => "Zero"
    case list: List[_] if list.isEmpty => "Empty list"
    case list: List[_] => s"List with ${list.length} elements"
    case Some(value) => s"Option with value: $value"
    case None => "Empty option"
    case _ => "Unknown type"
  }
  
  def processShape(shape: Shape): String = shape match {
    case Circle(r) if r > 10 => s"Large circle with area ${shape.area:.2f}"
    case Circle(r) => s"Small circle with area ${shape.area:.2f}"
    case Rectangle(w, h) if w == h => s"Square with area ${shape.area:.2f}"
    case Rectangle(w, h) => s"Rectangle ${w}x${h} with area ${shape.area:.2f}"
    case Triangle(b, h, _, _) => s"Triangle with base $b, height $h, area ${shape.area:.2f}"
  }
}

// Main application object
object ScalaSampleApp extends App {
  println("=== Scala Sample Program ===")
  println()
  
  // Object-oriented programming
  println("--- Object-Oriented Programming ---")
  val dog = new Dog("Buddy", "Golden Retriever")
  val cat = new Cat("Whiskers", true)
  
  println(dog.description)
  println(s"Dog says: ${dog.makeSound}")
  println(dog.fetch("ball"))
  println()
  
  println(cat.description)
  println(s"Cat says: ${cat.makeSound}")
  println(cat.purr())
  println()
  
  // Case classes and immutable data
  println("--- Case Classes and Immutable Data ---")
  val person1 = Person("John", "Doe", 30, "john.doe@example.com")
  val person2 = Person.fromEmail("jane.smith@example.com", 25)
  
  println(s"Person 1: ${person1.fullName}, Age: ${person1.age}, Adult: ${person1.isAdult}")
  println(s"Person 2: ${person2.fullName}, Age: ${person2.age}, Adult: ${person2.isAdult}")
  
  val olderPerson1 = person1.withAge(31)
  println(s"Person 1 after birthday: ${olderPerson1.fullName}, Age: ${olderPerson1.age}")
  println()
  
  // Collections and functional programming
  println("--- Collections and Functional Programming ---")
  val numbers = List(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
  
  val evenNumbers = numbers.filter(_ % 2 == 0)
  val squaredNumbers = numbers.map(x => x * x)
  val sum = numbers.reduce(_ + _)
  val product = numbers.filter(_ <= 5).product
  
  println(s"Original numbers: $numbers")
  println(s"Even numbers: $evenNumbers")
  println(s"Squared numbers: $squaredNumbers")
  println(s"Sum: $sum")
  println(s"Product of numbers <= 5: $product")
  
  // Higher-order functions
  val names = List("alice", "bob", "charlie", "diana")
  val capitalizedNames = names.map(_.capitalize)
  val longNames = names.filter(_.length > 4)
  val namesByLength = names.groupBy(_.length)
  
  println(s"Capitalized names: $capitalizedNames")
  println(s"Long names: $longNames")
  println(s"Names by length: $namesByLength")
  println()
  
  // Pattern matching
  println("--- Pattern Matching ---")
  val shapes = List(
    Circle(5.0),
    Rectangle(4.0, 6.0),
    Rectangle(5.0, 5.0),
    Triangle(3.0, 4.0, 5.0, 4.0)
  )
  
  shapes.foreach { shape =>
    println(PatternMatchingExamples.processShape(shape))
  }
  println()
  
  val values: List[Any] = List("Hello", 42, -5, 0, List(1, 2, 3), List(), Some("test"), None)
  values.foreach { value =>
    println(PatternMatchingExamples.describeValue(value))
  }
  println()
  
  // Math utilities
  println("--- Mathematical Operations ---")
  println(s"Fibonacci(10): ${MathUtils.fibonacci(10)}")
  println(s"Factorial(8): ${MathUtils.factorial(8)}")
  println(s"Prime numbers from 1 to 20: ${(1 to 20).filter(MathUtils.isPrime)}")
  println()
  
  // Generic containers
  println("--- Generic Programming ---")
  val intContainer = new GenericContainer(42)
  val stringContainer = intContainer.transform(_.toString)
  val doubleContainer = intContainer.transform(_.toDouble)
  
  println(s"Int container: $intContainer")
  println(s"String container: $stringContainer")
  println(s"Double container: $doubleContainer")
  println()
  
  // Option and Either examples
  println("--- Option and Either ---")
  val optionalValue: Option[String] = Some("Hello, Scala!")
  val emptyOption: Option[String] = None
  
  println(s"Optional value: ${optionalValue.getOrElse("Default")}")
  println(s"Empty option: ${emptyOption.getOrElse("Default")}")
  
  val result: Either[String, Int] = Try(42 / 2).map(Right(_)).getOrElse(Left("Division error"))
  result match {
    case Right(value) => println(s"Success: $value")
    case Left(error) => println(s"Error: $error")
  }
  println()
  
  // For comprehensions
  println("--- For Comprehensions ---")
  val pairs = for {
    x <- 1 to 3
    y <- 1 to 3
    if x != y
  } yield (x, y)
  
  println(s"Pairs where x != y: $pairs")
  
  val nestedList = List(List(1, 2), List(3, 4), List(5, 6))
  val flattened = for {
    sublist <- nestedList
    element <- sublist
  } yield element * 2
  
  println(s"Flattened and doubled: $flattened")
  println()
  
  // Async operations (simplified for demo)
  println("--- Asynchronous Operations ---")
  val futureResult = AsyncOperations.processDataAsync(List(1, 2, 5, 8, 12, 15))
  
  futureResult.onComplete {
    case Success(result) => println(s"Async processing result: $result")
    case Failure(exception) => println(s"Async processing failed: ${exception.getMessage}")
  }
  
  // Wait a bit for async operation to complete
  Thread.sleep(100)
  
  println()
  println("Program completed successfully!")
} 